### PR TITLE
Delays Notifications

### PR DESCRIPTION
**Pull request by cindy-x-liang**
Will delete my print statements and change the URL to use the actual backend when everything is ready to go :D 

## Overview

Additing changes to livetracking.py to handle sending delay notifications to a user. The microservice will check every 30 seconds if any of the tripIDs in the notif_requests dictionary is delayed, meaning it shows up in the RTF data. If so, it sends a notification to the main ithaca-transit-backend to all the device tokens that are values to that tripID key. 

## Changes Made

When a user requests to be notified of delays for a given trip, their device token will be added to the key that corresponds to that tripID in notif_requests. Every 30 seconds, the RTF data is iterated through to see if any of the tripIDs correspond to tripIDs is notif_requests, meaning that the trip is now delayed. If so, we send a notification to all users in that tripID. We use a json file to store notif_requests data.

### CODE REVIEW

File: app.py [hunk 1]
```diff
@@ -25,6 +26,7 @@ def get_gtfs():
 @app.route("/")
 @app.route("/rtf")
 def get_rtf():
+    print('bye2')
```
Comment (Aayush-Agnihotri) [inline app.py:29]: print("bye3") 😼

File: app.py [hunk 2]
```diff
@@ -42,6 +44,19 @@ def get_gtfs_date():
 def get_vehicles():
     return jsonify(get_vehicles_data())
 
+@app.route("/delayNotifs/", methods=["POST"])
+def get_delayNotifs():
+    #notif_requests = {'t65B-b1F42-slC':'hi'}
+    body = json.loads(request.data)
+    trip = body.get("tripId")
+    print(body)
```
Comment (Aayush-Agnihotri) [inline app.py:52]: print("bye body")

File: app.py [hunk 3]
```diff
@@ -50,12 +65,14 @@ def get_vehicles():
     fetch_rtf(rtf_event)
     fetch_stops(stops_event)
     fetch_vehicles(vehicles_event)
+    start_notif_timer()
     time.sleep(1)
-    app.run(host="0.0.0.0", port=5000)
+    app.run(host="0.0.0.0", port=8000)
```
Comment (Aayush-Agnihotri) [inline app.py:70]: wait we should leave the port at 5000 right?

File: src/live_tracking.py [hunk 4]
```diff
@@ -38,6 +46,53 @@ def fetch_rtf(event):
...
+def add_delay(trip,deviceToken):
+    global notif_requests
+    # notif_requests = fetch_requests()
+    if trip in notif_requests:
+        notif_requests[trip].append(deviceToken)
+    else:
+        notif_requests[trip] = [deviceToken]
+
+    save_notifs(notif_requests)
+    send_notifs()
```
Comment (Aayush-Agnihotri) [inline src/live_tracking.py:60]: wait we wanna send notifs here or have send_notifs running on like a schedule to make sure users are notified asap?

File: src/live_tracking.py [hunk 5]
```diff
@@ -38,6 +46,53 @@ def fetch_rtf(event):
...
+                print(user)
+                #sends a notification to a device if it is waiting for a delay
+                #notification
+                send_notif({'deviceToken':user, 'routeID':rtf_data[id]['routeId']})
+            #deletes this tripId from notif_requests as it is no longer needed
+            print('here')
+            del notif_requests[id]
+            
+    print(notif_requests)
+    print('yay')
```
Comment (Aayush-Agnihotri) [inline src/live_tracking.py:78]: YAY!!

File: src/live_tracking.py [hunk 6]
```diff
@@ -38,6 +46,53 @@ def fetch_rtf(event):
...
+def start_notif_timer():
+    threading.Timer(30, send_notifs).start()
+
+def send_notif(data):
+    url = 'http://127.0.0.1:3000/api/v1/microserviceNotif'
+    headers = { 'Content-Type': 'application/json' }
+    response = requests.post(url, json=data, headers=headers)
+    print(response)
+    print(data)
+    print('hi')
```
Comment (Aayush-Agnihotri) [inline src/live_tracking.py:94]: hello

**Review by Aayush-Agnihotri — APPROVED** [review]
LGTM

**Review by nfq2 — APPROVED** [review]

File: src/live_tracking.py (hunk 4 above)
Comment (cindy-x-liang) [reply src/live_tracking.py:60]: True the function call here isn't necessary since send_notifs() is already being scheduled to run every 30 seconds

**Pull request merged**